### ISSUE TITLE
ENH: Make `np.object_` generic for better object array type hinting

### ISSUE THREAD

**Issue by flexatone** [numpy/numpy#25351]
### Proposed new feature or change:

Greetings!

Could `np.object_` be made generic? 

For example, if I have an object array of two-element integer tuples, it would be ideal if I could type hint the array with `np.dtype[np.object_[tuple[int, int]]]`; this would allow us to distinguish between object arrays by their contained types. (I.e., `np.dtype[np.object_[tuple[int, int]]]` is not the same as `np.dtype[np.object_[datetime.date]]`.)

I am happy to explore a PR if this seems like a good enhancement.

I posted this on the "numpy-discussion" forum but there was no feedback.

**Comment by basnijholt** [comment]
This sound like a good idea and solve the problem I am running [into](https://discuss.python.org/t/generics-in-metadata-of-annotated/62059).

Might also address #24738.

**Comment by JaRoSchm** [comment]
This comment in the numpy stubs may be related, right?
https://github.com/numpy/numpy/blob/0dbee7ec68f1adcabfb0344c53d65b7f385ae6b5/numpy/__init__.pyi#L3969-L3972
Is https://github.com/python/mypy/issues/15182 this long-standing bug?

**Comment by jorenham** [comment]
> This comment in the numpy stubs may be related, right?
> 
> [numpy/numpy/__init__.pyi](https://github.com/numpy/numpy/blob/0dbee7ec68f1adcabfb0344c53d65b7f385ae6b5/numpy/__init__.pyi#L3969-L3972)
> 
> Lines 3969 to 3972 in [0dbee7e](/numpy/numpy/commit/0dbee7ec68f1adcabfb0344c53d65b7f385ae6b5)
> 
>  # NOTE: Because mypy has some long-standing bugs related to `__new__`, `object_` can't 
>  # be made generic. 
>  @final 
>  class object_(_RealMixin, generic): 
> 
> Is [python/mypy#15182](https://github.com/python/mypy/issues/15182) this long-standing bug (taken from [python/mypy#18343 (comment)](https://github.com/python/mypy/issues/18343#issuecomment-2571784915))?

Yes that's right. A while back I attempted to make `object_`, but it broke mypy in many unexpected ways, so I unfortunately had to revert it.

**Comment by jorenham** [comment]
> Yes that's right. A while back I attempted to make `object_` generic, but it broke mypy in many unexpected ways, so I unfortunately had to revert it.
> 
> So in other words, `object_` cannot be made generic until [python/mypy#15182](https://github.com/python/mypy/issues/15182) is fixed.

https://github.com/python/mypy/issues/15182 has been fixed in mypy 2.2.0, so this should now be possible, and I'm planning to give it another shot soon.

But for bookkeeping purposes I'm going to close this, because it turns out that this is a duplicate of #24738

**Comment by flexatone** [comment]
@jorenham : thanks, I look forward to being able to use this in a future release!

**Comment by jorenham** [comment]
I opened a PR: https://github.com/numpy/numpy/pull/32075. It would help a lot if you could take a look and see if this will fix your use case :)

**Comment by flexatone** [comment]
> I opened a PR: [#32075](https://github.com/numpy/numpy/pull/32075). It would help a lot if you could take a look and see if this will fix your use case :)

@jorenham : yes, that is exactly what I was hoping for. Thanks!
